sdcm/kafka/kafka_producer.py

```diff
@@ -0,0 +1,464 @@
+        super().__init__(tester, params, timeout, **kwargs)
+        self.delivered_count = 0
+        self.undelivered_count = 0
+        self.current_key = 1
+
+        self.record_size = int(self.kafka_options.get('record-size', 1024))
+
+        producer_properties = {}
+        if 'producer-property' in self.kafka_options:
+            if isinstance(self.kafka_options['producer-property'], dict):
+                producer_properties.update(self.kafka_options['producer-property'])
+            elif isinstance(self.kafka_options['producer-property'], list):
+                for prop in self.kafka_options['producer-property']:
+                    producer_properties.update(prop)
+        if 'property' in self.kafka_options:
+            if isinstance(self.kafka_options['property'], dict):
+                producer_properties.update(self.kafka_options['property'])
+            elif isinstance(self.kafka_options['property'], list):
+                for prop in self.kafka_options['property']:
+                    producer_properties.update(prop)
+
+        producer_config = {
+            'bootstrap.servers': ','.join(self.kafka_addresses),
+        }
+        # The following parameters can be used to tune/optimize performance and reliability of the producer
+        # producer_config['linger.ms'] = int(producer_properties.get('linger.ms', kwargs.get('linger_ms', 5)))
+        # producer_config['batch.num.messages'] = int(
+        #     producer_properties.get('batch.num.messages', kwargs.get('batch_num_messages', 10000)))
+        # producer_config['acks'] = producer_properties.get('acks', kwargs.get('acks', '1'))
+        # producer_config['compression.type'] = producer_properties.get(
+        #     'compression.type', kwargs.get('compression_type', 'none'))
+
+        producer_config.update(producer_properties)
+
+        self.producer = Producer(producer_config)
+        self.wait_for_topic(self.topic, timeout=60)
+
+    def get_topics(self) -> list[str]:
+        admin_client = Producer({'bootstrap.servers': ','.join(self.kafka_addresses)})
+        topics = admin_client.list_topics(timeout=10).topics.keys()
+        LOGGER.debug("Available topics: %s", topics)
+        return topics
+
+    def wait_for_topic(self, topic: str, timeout: int) -> None:
+        wait_for(lambda: topic in self.get_topics(), text=f"waiting for topic={topic}", timeout=timeout)
+
+    def delivery_report(self, err: str, record: Message) -> None:
+        if err:
+            self.undelivered_count += 1
+            LOGGER.error("Kafka record delivery failed: %s", err)
+        else:
+            self.delivered_count += 1
+            LOGGER.debug(
+                "Kafka record delivered to %s [%s] at offset %s", record.topic(), record.partition(), record.offset())
+
+    def run(self) -> None:
+        start_time = time.time()
+        while not self.termination_event.is_set():
+            if self.delivered_count + self.undelivered_count >= self.num_records:
```

Comment:
btw. why we cannot rely on `current_key` number?

Reply:
We can. Initial implementation was not having pagination/batching, so the thread instance didn't have current_key attribute at that time.

Reply:
After a second thought I think it's better to keep this from of a condition. It keeps semantics consistent - the same objects are checked (requested num of records in kafka against currently processed nu of records).
current_key is a field of a Kafka message, and reading the condition when using it would mean - compare value of a field of kafka message with the requested num of records.